Firestore/core/src/firebase/firestore/remote/remote_store.h

```diff
@@ -0,0 +1,216 @@
+#include "Firestore/core/src/firebase/firestore/remote/watch_stream.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+@class FSTLocalStore;
+@class FSTMutationBatchResult;
+@class FSTQueryData;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * A protocol that describes the actions the FSTRemoteStore needs to perform on
+ * a cooperating synchronization engine.
+ */
+@protocol FSTRemoteSyncer
+
+/**
+ * Applies one remote event to the sync engine, notifying any views of the
+ * changes, and releasing any pending mutation batches that would become visible
+ * because of the snapshot version the remote event contains.
+ */
+- (void)applyRemoteEvent:
+    (const firebase::firestore::remote::RemoteEvent&)remoteEvent;
+
+/**
+ * Rejects the listen for the given targetID. This can be triggered by the
+ * backend for any active target.
+ *
+ * @param targetID The targetID corresponding to a listen initiated via
+ *     -listenToTargetWithQueryData: on FSTRemoteStore.
+ * @param error A description of the condition that has forced the rejection.
+ * Nearly always this will be an indication that the user is no longer
+ * authorized to see the data matching the target.
+ */
+- (void)rejectListenWithTargetID:
+            (const firebase::firestore::model::TargetId)targetID
+                           error:
+                               (NSError*)error;  // NOLINT(readability/casting)
+
+/**
+ * Applies the result of a successful write of a mutation batch to the sync
+ * engine, emitting snapshots in any views that the mutation applies to, and
+ * removing the batch from the mutation queue.
+ */
+- (void)applySuccessfulWriteWithResult:
+    (FSTMutationBatchResult*)batchResult;  // NOLINT(readability/casting)
+
+/**
+ * Rejects the batch, removing the batch from the mutation queue, recomputing
+ * the local view of any documents affected by the batch and then, emitting
+ * snapshots with the reverted value.
+ */
+- (void)
+    rejectFailedWriteWithBatchID:(firebase::firestore::model::BatchId)batchID
+                           error:
+                               (NSError*)error;  // NOLINT(readability/casting)
+
+/**
+ * Returns the set of remote document keys for the given target ID. This list
+ * includes the documents that were assigned to the target when we received the
+ * last snapshot.
+ */
+- (firebase::firestore::model::DocumentKeySet)remoteKeysForTarget:
+    (firebase::firestore::model::TargetId)targetId;
+
+@end
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+class RemoteStore : public TargetMetadataProvider, public WatchStreamCallback {
+ public:
+  RemoteStore(FSTLocalStore* local_store,
+              Datastore* datastore,
+              util::AsyncQueue* worker_queue,
+              std::function<void(model::OnlineState)> online_state_handler);
+
+  // TODO(varconst): remove the getters and setters
+  id<FSTRemoteSyncer> sync_engine() {
```

Comment:
These are only necessary for the short transitional period, until the rest of `FSTRemoteStore` is migrated.